actors/migration/top.go

```diff
@@ -75,41 +81,65 @@ var migrations = map[cid.Cid]ActorMigration{ // nolint:varcheck,deadcode,unused
 	},
 }
 
-// A phoenix tracks the unburning of funds
-type phoenix struct {
-	burntBalance abi.TokenAmount
-}
+// TODOS
+func migrateOneActor(ctx context.Context, store cbor.IpldStore, addr address.Address, actorIn *states.Actor, actorsOut *states.Tree, transferCh chan big.Int, errCh chan error) {
+	migration := migrations[actorIn.Code]
 
-func (p *phoenix) load(ctx context.Context, actorsIn *states0.Tree) error {
-	burntFundsActor, found, err := actorsIn.GetActor(builtin0.BurntFundsActorAddr)
-	if err != nil {
-		return err
+	// This will be migrated at the end
+	if actorIn.Code == builtin0.VerifiedRegistryActorCodeID {
+		sem.Release(1)
+		return
 	}
-	if !found {
-		return xerrors.Errorf("burnt funds actor not in tree")
+	if actorIn.Code == builtin0.StorageMinerActorCodeID {
+		// setup migration fields
+		mm := migration.StateMigration.(*minerMigrator)
+		// TODO this is broken
+		// Miner migration should have a different signature that
+		// takes balance as input and returns the amount that we
+		// need to transfer to the miner.  Can't rely on shared state
+		// of mm.MinerBalance as its not thread safe and don't want to introduce
+		// contention by locking it.
+		mm.MinerBalance = big.Zero()
 	}
-	p.burntBalance = burntFundsActor.Balance
-	return nil
-}
-
-func (p *phoenix) transfer(amt abi.TokenAmount) error {
-	p.burntBalance = big.Sub(p.burntBalance, amt)
-	if p.burntBalance.LessThan(big.Zero()) {
-		return xerrors.Errorf("migration programmer error burnt funds balance falls to %v, below zero", p.burntBalance)
+	headOut, err := migration.StateMigration.MigrateState(ctx, store, actorIn.Head)
+	if err != nil {
+		err = xerrors.Errorf("state migration error on %s actor at addr %s: %w", builtin.ActorNameByCode(migration.OutCodeCID), addr, err)
+		sem.Release(1)
```

Comment:
preventing deadlock in the rare case all worker goroutines are sending an error on channel and the main thread producing work is blocking on the semaphore.  This makes it so that there could be an extra worker thread but that seems like nbd, not being able to cleanly defer seems like more of a problem.

Reply:
Add that motivation to the code, not just the PR.